src/valid/handles.rs

```diff
@@ -0,0 +1,637 @@
+            &crate::Statement::Store { pointer, value } => {
+                validate_expr(pointer)?;
+                validate_expr(value)?;
+                Ok(())
+            }
+            &crate::Statement::ImageStore {
+                image,
+                coordinate,
+                array_index,
+                value,
+            } => {
+                validate_expr(image)?;
+                validate_expr(coordinate)?;
+                validate_expr_opt(array_index)?;
+                validate_expr(value)?;
+                Ok(())
+            }
+            &crate::Statement::Atomic {
+                pointer,
+                fun,
+                value,
+                result,
+            } => {
+                validate_expr(pointer)?;
+                match fun {
+                    crate::AtomicFunction::Add
+                    | crate::AtomicFunction::Subtract
+                    | crate::AtomicFunction::And
+                    | crate::AtomicFunction::ExclusiveOr
+                    | crate::AtomicFunction::InclusiveOr
+                    | crate::AtomicFunction::Min
+                    | crate::AtomicFunction::Max => (),
+                    crate::AtomicFunction::Exchange { compare } => validate_expr_opt(compare)?,
+                };
+                validate_expr(value)?;
+                validate_expr(result)?;
+                Ok(())
+            }
+            &crate::Statement::Call {
+                function,
+                ref arguments,
+                result,
+            } => {
+                Self::validate_function_handle(function, functions)?;
+                for arg in arguments.iter().copied() {
+                    validate_expr(arg)?;
+                }
+                validate_expr_opt(result)?;
+                Ok(())
+            }
+            &crate::Statement::Break
+            | &crate::Statement::Continue
+            | &crate::Statement::Kill
+            | &crate::Statement::Barrier(_) => Ok(()),
+        })
+    }
+}
+
+#[derive(Clone, Debug)]
+pub(super) enum HandleValidationError {
+    InvalidHandle(InvalidHandleError),
+    InvalidArraySize(InvalidArraySizeError),
+}
+
+impl From<InvalidHandleError> for HandleValidationError {
+    fn from(source: InvalidHandleError) -> Self {
+        Self::InvalidHandle(source)
+    }
+}
+
+impl From<BadHandle> for HandleValidationError {
+    fn from(source: BadHandle) -> Self {
+        Self::InvalidHandle(source.into())
+    }
+}
+
+impl From<FwdDepError> for HandleValidationError {
+    fn from(source: FwdDepError) -> Self {
+        Self::InvalidHandle(source.into())
+    }
+}
+
+impl From<BadRangeError> for HandleValidationError {
+    fn from(source: BadRangeError) -> Self {
+        Self::InvalidHandle(source.into())
+    }
+}
+
+impl From<InvalidArraySizeError> for HandleValidationError {
+    fn from(source: InvalidArraySizeError) -> Self {
+        Self::InvalidArraySize(source)
+    }
+}
```

Comment:
N.B. that the `From` conversions here may not be necessary with some well-placed `map_err`; I hadn't tried that, but that could cut the trait implementations down, in case that changes your calculus.

WRT moving to `ValidationError`, I'm not feeling picky, I guess. Moved to "just" return a `ValidationError` with df755c36.